src/ProxyManager/ProxyGenerator/AccessInterceptorScopeLocalizer/MethodGenerator/BindProxyProperties.php

```diff
@@ -70,10 +65,101 @@ public function __construct(
                 . ')->__invoke();';
         }
 
-        $this->setBody(
-            ($localizedProperties ? implode("\n\n", $localizedProperties) . "\n\n" : '')
-            . '$this->' . $prefixInterceptors->getName() . " = \$prefixInterceptors;\n"
-            . '$this->' . $suffixInterceptors->getName() . ' = $suffixInterceptors;'
+        $bodyLines = [
+            implode("\n\n", $localizedProperties),
+        ];
+
+        $nonReferenceableProperties = Properties::fromReflectionClass($originalClass)
+            ->onlyNonReferenceableProperties()
+            ->onlyInstanceProperties();
+
+        if (! $nonReferenceableProperties->empty()) {
+            $bodyLines[] = '$class = new \ReflectionObject($localizedObject);';
+            $bodyLines[] = $this->generateClassMapInitializationCode($nonReferenceableProperties);
+
+            foreach ($nonReferenceableProperties->getInstanceProperties() as $property) {
+                $bodyLines[] = $this->generateCodeForPropertyBinding($property);
+            }
+        }
+
+        $bodyLines[] = sprintf('$this->%s = $prefixInterceptors;', $prefixInterceptors->getName());
+        $bodyLines[] = sprintf('$this->%s = $suffixInterceptors;', $suffixInterceptors->getName());
+        $this->setBody(implode(PHP_EOL, $bodyLines));
+    }
+
+    private function generateClassMapInitializationCode(Properties $properties): string
+    {
+        $classMapInitializationCode = '$classesMap = []; // Class name to ReflectionClass map' . PHP_EOL;
+
+        $mappedClasses = [];
+        foreach ($properties->getPrivateProperties() as $property) {
+            $declaringClassName = $property->getDeclaringClass()->getName();
+            if (array_key_exists($declaringClassName, $mappedClasses)) {
+                continue;
+            }
+
+            $classMapInitializationCode .= sprintf(
+                '$classesMap[\'%s\'] = new \ReflectionClass(\'%s\');',
+                $declaringClassName,
+                '\\' . $declaringClassName
+            );
+            $classMapInitializationCode .= PHP_EOL;
+
+            $mappedClasses[$declaringClassName] = null;
+        }
+
+        return $classMapInitializationCode;
+    }
+
+    private function generateCodeForPropertyBinding(ReflectionProperty $property): string
+    {
+        $declaringClassName     = $property->getDeclaringClass()->getName();
+        $localizedPropertyCode  = $property->isPrivate()
+            ? sprintf(
+                '$property = $classesMap[\'%s\']->getProperty(\'%s\');',
+                $declaringClassName,
+                $property->getName()
+            )
+            : sprintf('$property = $class->getProperty(\'%s\');', $property->getName());
+        $localizedPropertyCode .= PHP_EOL;
+
+        if (! $property->isPublic()) {
+            $localizedPropertyCode .= '$property->setAccessible(true);' . PHP_EOL;
```

Comment:
can be removed since https://wiki.php.net/rfc/make-reflection-setaccessible-no-op